src/Lecture2.hs

```diff
@@ -150,14 +170,219 @@ You're free to define any helper functions.
+attackToInt (Attack atk) = atk
+
+instance Show Gold where
+    show (Gold g) = "💰 " ++ show g
+
+instance Show XP where
+    show (XP xp) = "✨ " ++ show xp
+
+instance Show Health where
+    show (Health h) = "💖 " ++ show h
+
+instance Show Attack where
+    show (Attack atk) = "🔪 " ++ show atk
+
+instance Show Endurance where
+    show (Endurance e) = "✌ " ++ show e
+
+data Treasure = Treasure
+
+class ChestT a where
+    chestGold :: a -> Gold
+    chestTreasure :: a -> Maybe Treasure
+
+newtype ChestWithTreasureT = ChestWithTreasureT (Gold, Treasure)
+
+instance ChestT ChestWithTreasureT where
+    chestGold (ChestWithTreasureT (g, _)) = g
+    chestTreasure (ChestWithTreasureT (_, t)) = Just t
+
+newtype AcidChestT = AcidChestT Gold
+
+instance ChestT AcidChestT where
+    chestGold (AcidChestT g) = g
+    chestTreasure _ = Nothing
+
+data Chest = ChestWithTreasure ChestWithTreasureT
+           | AcidChest AcidChestT
+
+instance Show Chest where
+    show chest = case chestTreasure chest of
+        Just _ -> "🧰 (" ++ show (chestGold chest) ++ ", with treasure)"
+        Nothing -> "🧰 (" ++ show (chestGold chest) ++ ", no treasure)"
+
+instance ChestT Chest where
+    chestGold (ChestWithTreasure cwt) = chestGold cwt
+    chestGold (AcidChest ac) = chestGold ac
+    chestTreasure (ChestWithTreasure cwt) = chestTreasure cwt
+    chestTreasure (AcidChest ac) = chestTreasure ac
+
+treasureChest :: Gold -> Treasure -> Chest
+treasureChest g t = ChestWithTreasure (ChestWithTreasureT (g, t))
+
+acidChest :: Gold -> Chest
+acidChest g = AcidChest (AcidChestT g)
+
+class DragonT a where
+    dragonExperience :: a -> XP
+    dragonFirePower :: a -> Attack
+    dragonHealth :: a -> Health
+    dragonSlayCounter :: a -> Int
+    dragonChest :: a -> Chest
+
+data RedDragonT = RedDragonT
+    { redDragonHealth :: Health
+    , redDragonFirePower :: Attack
+    , redDragonSlayCounter :: Int
+    , redDragonGold :: Gold
+    , redDragonTreasure :: Treasure
+    }
+
+data BlackDragonT = BlackDragonT
+    { blackDragonHealth :: Health
+    , blackDragonFirePower :: Attack
+    , blackDragonSlayCounter :: Int
+    , blackDragonGold :: Gold
+    , blackDragonTreasure :: Treasure
+    }
+
+data GreenDragonT = GreenDragonT
+    { greenDragonHealth :: Health
+    , greenDragonFirePower :: Attack
+    , greenDragonSlayCounter :: Int
+    , greenDragonGold :: Gold
+    }
+
+instance DragonT RedDragonT where
+    dragonExperience _ = XP 100
+    dragonFirePower = redDragonFirePower
+    dragonHealth = redDragonHealth
+    dragonSlayCounter = redDragonSlayCounter
+    dragonChest d = treasureChest (redDragonGold d) (redDragonTreasure d)
+
+instance DragonT BlackDragonT where
+    dragonExperience _ = XP 150
+    dragonFirePower = blackDragonFirePower
+    dragonHealth = blackDragonHealth
+    dragonSlayCounter = blackDragonSlayCounter
+    dragonChest d = treasureChest (blackDragonGold d) (blackDragonTreasure d)
+
+instance DragonT GreenDragonT where
```

Comment:
That's a lot of typeclasses and instances 😮 

I guess, you've optimized your solution for extensibility and the open-world assumption 🙂

Reply:
actually, I tried not using custom tyoeclasses at all, but I just couldn't. I guess I'm used to the Scala give-evetything-a-trait approach